src/main.ts

```diff
@@ -649,38 +652,48 @@ function generateEncode(ctx: Context, fullName: string, messageDesc: DescriptorP
     const fieldName = maybeSnakeToCamel(field.name, options);
 
     // get a generic writer.doSomething based on the basic type
-    let writeSnippet: (place: string) => Code;
-    if (isPrimitive(field)) {
+    let writeSnippet: (place: string, forceDefaultSerialization: boolean) => Code;
+    if (isScalar(field) || isEnum(field)) {
       const tag = ((field.number << 3) | basicWireType(field.type)) >>> 0;
-      writeSnippet = (place) => code`writer.uint32(${tag}).${toReaderCall(field)}(${place})`;
+      writeSnippet = (place, _) => code`writer.uint32(${tag}).${toReaderCall(field)}(${place})`;
     } else if (isTimestamp(field)) {
       const tag = ((field.number << 3) | 2) >>> 0;
       const type = basicTypeName(ctx, field, { keepValueType: true });
-      writeSnippet = (place) =>
-        code`${type}.encode(${utils.toTimestamp}(${place}), writer.uint32(${tag}).fork()).ldelim()`;
+      writeSnippet = (place, forceDefaultSerialization) =>
+        code`${type}.encode(${utils.toTimestamp}(${place}), writer.uint32(${tag}).fork(), ${forceDefaultSerialization}).ldelim()`;
     } else if (isValueType(ctx, field)) {
       const tag = ((field.number << 3) | 2) >>> 0;
       const type = basicTypeName(ctx, field, { keepValueType: true });
-      writeSnippet = (place) => code`${type}.encode({ value: ${place}! }, writer.uint32(${tag}).fork()).ldelim()`;
+      writeSnippet = (place, forceDefaultSerialization) =>
+        code`${type}.encode({ value: ${place}! }, writer.uint32(${tag}).fork(), ${forceDefaultSerialization}).ldelim()`;
     } else if (isMessage(field)) {
       const tag = ((field.number << 3) | 2) >>> 0;
       const type = basicTypeName(ctx, field);
-      writeSnippet = (place) => code`${type}.encode(${place}, writer.uint32(${tag}).fork()).ldelim()`;
+      writeSnippet = (place, forceDefaultSerialization) =>
+        code`${type}.encode(${place}, writer.uint32(${tag}).fork(), ${forceDefaultSerialization}).ldelim()`;
     } else {
       throw new Error(`Unhandled field ${field}`);
     }
 
     if (isRepeated(field)) {
       if (isMapType(ctx, messageDesc, field)) {
+        // protobuf.js does not support map entries with keys set and values unset.
```

Comment:
Hm, just curious, but how does protobuf.js not support unset values? I.e. what would break if we did serialize this with a key set but value unset?

Reply:
I'm glad you ask. It felt like a bug in pbjs to me. I tried to reproduce it and it turns out I cannot reproduce it anymore. Maybe outdated pbjs code during development 🤷‍♂️

I was able to remove all the special case handling, making all of this much nicer.